### PR TITLE
Redo field updates for faster compile

### PR DESCRIPTION
**Pull request by TimSheard**
Field updates like
```haskell
 foo {field1 = exp }
```

Seem to take forever to compile if  'exp' has a complicated type.
Empirical analysis shows that replacing them with simple update functions can dramatically speed things up
for example.
```haskell
update1 f = \ x -> (x{field1 = f (field1 x)})
```

we recombined  `Test.Cardano.Ledger.Generic.GenState.Types`, `Test.Cardano.Ledger.Generic.GenState.GenScript`, and `Test.Cardano.Ledger.Generic.GenState.Generate` into `Test.Cardano.Ledger.Generic.GenState`

we also redid `Test.Cardano.Ledger.Generic.TxGen`

cardano-ledger-test now cuts approximately 20 minutes from compile time.

### CODE REVIEW

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs [hunk 1]
```diff
@@ -13,7 +13,9 @@
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE UndecidableInstances #-}
 
-module Test.Cardano.Ledger.Generic.TxGen
+module Test.Cardano.Ledger.Generic.TxGen where
+
+{-
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Explicit exports got commented out

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs [hunk 2]
```diff
@@ -588,47 +605,48 @@ genDCerts proof slot = do
...
-                        insertIfNotPresent dcs (Map.delete deregCred regCreds) Nothing
-                          <$> lookupPlutusScript deregCred Cert
-                  Just (Coin _) -> pure (dcs, ss, regCreds)
-            -- Either Reward balance is not zero, or no Credential, so skip making certificate
-            | Delegate (Delegation delegCred delegKey) <- d ->
-                let (dcs', regCreds')
-                      | delegCred `Map.member` regCreds = (dcs, regCreds)
-                      | otherwise -- In order to Delegate, the delegCred must exist in rewards.
+          DCertDeleg d -> case d of
+            (RegKey regCred) ->
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Redundant parenthesis:
```suggestion
            RegKey regCred ->
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs [hunk 3]
```diff
@@ -588,47 +605,48 @@ genDCerts proof slot = do
...
+              if regCred `Map.member` regCreds -- Can't register if it is already registered
+                then pure (dcs, ss, regCreds)
+                else pure (dc : dcs, ss, Map.insert regCred (Coin 99) regCreds) -- 99 is a NonZero Value
+            (DeRegKey deregCred) ->
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Redundant parenthesis:
```suggestion
            DeRegKey deregCred ->
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs [hunk 4]
```diff
@@ -588,47 +605,48 @@ genDCerts proof slot = do
...
+                -- No credential, skip making certificate
+                Just (Coin 0) ->
+                  -- Ok to make certificate, rewards balance is 0
+                  if Set.member deregCred honest
+                    then pure (dcs, ss, regCreds)
+                    else
+                      insertIfNotPresent dcs (Map.delete deregCred regCreds) Nothing
+                        <$> lookupPlutusScript deregCred Cert
+                Just (Coin _) -> pure (dcs, ss, regCreds)
+            (Delegate (Delegation delegCred delegKey)) ->
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Redundant parenthesis:
```suggestion
            Delegate (Delegation delegCred delegKey) ->
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs [hunk 5]
```diff
@@ -588,47 +605,48 @@ genDCerts proof slot = do
...
-                  Nothing -> pure (dcs, ss, regCreds)
+
+                        ( DCertDeleg (RegKey delegCred) : dcs,
+                          Map.insert delegCred (Coin 99) regCreds
+                        )
+               in insertIfNotPresent dcs' regCreds' (Just delegKey)
+                    <$> lookupPlutusScript delegCred Cert
+          DCertPool d -> case d of
+            (RegPool _) -> pure (dc : dcs, ss, regCreds)
+            (RetirePool kh _) -> do
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Redundant parenthesis:
```suggestion
            RegPool _ -> pure (dc : dcs, ss, regCreds)
            RetirePool kh _ -> do
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs [hunk 6]
```diff
@@ -987,30 +1009,14 @@ genAlonzoTxAndInfo proof slot = do
             AuxData' []
           ]
   count <- gets (mCount . gsModel)
+  modifyGenStateInitialRewards (\x -> Map.unions [x, newRewards])
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Only two maps:
```suggestion
  modifyGenStateInitialRewards (\x -> Map.union x newRewards)
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs [hunk 7]
```diff
@@ -199,6 +205,50 @@ lookupScript scriptHash mTag = do
...
+    Mary _ -> mkTimelockWit proof mTag script
+    Alonzo _ -> case script of
+      (TimelockScript timelock) -> mkTimelockWit proof mTag timelock
+      (PlutusScript _ _) -> pure (const [])
+    Babbage _ -> case script of
+      (TimelockScript timelock) -> mkTimelockWit proof mTag timelock
+      (PlutusScript _ _) -> pure (const [])
+    Conway _ -> case script of
+      (TimelockScript timelock) -> mkTimelockWit proof mTag timelock
+      (PlutusScript _ _) -> pure (const [])
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Love the switch to nested case statements. 

Slight improvement, redundant parenthesis:
```suggestion
      TimelockScript timelock -> mkTimelockWit proof mTag timelock
      PlutusScript _ _ -> pure (const [])
    Babbage _ -> case script of
      TimelockScript timelock -> mkTimelockWit proof mTag timelock
      PlutusScript _ _ -> pure (const [])
    Conway _ -> case script of
      TimelockScript timelock -> mkTimelockWit proof mTag timelock
      PlutusScript _ _ -> pure (const [])
```

**Review by lehins — APPROVED** [review]
I am glad that you found this weird compiler performance regression. Waiting for 5 minutes for a module to be compiled was a little too much.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs (hunk 1 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Yeah, that was a temporary change in order to compile it with bisection strategy to find what was taking forever. Forgot to undo it.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs (hunk 7 above)
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/TxGen.hs]: Changing to nested case statement had a small but not insignificant improvement in compile time. Makes it clearer too.

**Pull request merged**